### PR TITLE
Add std::source_location to ChipError for C++20 builds

### PR DESCRIPTION
**Pull request by mwswartwout**
Add std::source_location to ChipError when compiling with C++20.

This has the side effect of making the ChipError default constructor non-trivial, which made ChipDeviceEvent also not trivial. Switching ChipDeviceEvent construction to using the designated initializer syntax fixed this.

### CODE REVIEW

File: src/lib/core/CHIPError.h [hunk 1]
```diff
@@ -159,10 +183,17 @@ class ChipError
      * @note
      *  This is intended to be used only in foreign function interfaces.
      */
+#if __cplusplus >= 202002L
+    explicit constexpr ChipError(StorageType error) : mError(error) CHIP_INITIALIZE_ERROR_SOURCE(nullptr, 0, std::source_location()) {}
```
Comment (tcarmelveilleux) [inline src/lib/core/CHIPError.h:187]: Please add a unit test for this behavior in the chip error case, and include ifdef guards for c++20

File: src/lib/core/CHIPError.h [hunk 2]
```diff
@@ -123,23 +131,39 @@ class ChipError
      * @note
      *  The result is valid only if CanEncapsulate() is true.
      */
+#if __cplusplus >= 202002L
+    constexpr ChipError(Range range, ValueType value) :
+        mError(MakeInteger(range, (value & MakeMask(0, kValueLength)))) CHIP_INITIALIZE_ERROR_SOURCE(nullptr, 0, std::source_location())
```
Comment (bzbarsky-apple) [inline src/lib/core/CHIPError.h:136]: Won't this fail to compile if `__cplusplus >= 202002L` and `!CHIP_CONFIG_ERROR_SOURCE`?

File: src/lib/core/CHIPError.h [hunk 3]
```diff
@@ -123,23 +131,39 @@ class ChipError
...
+    {}
+    constexpr ChipError(Range range, ValueType value, const char * file, unsigned int line, std::source_location location = std::source_location::current()) :
```
Comment (bzbarsky-apple) [inline src/lib/core/CHIPError.h:138]: So how does this handle situations where the ChipError() constructor is called via a macro defined at location A but used at location B?  Which location ends up in `location`?

Comment (mwswartwout) [reply src/lib/core/CHIPError.h:138]: It should always show location B: https://godbolt.org/z/oThxMKKM4

Comment (bzbarsky-apple) [reply src/lib/core/CHIPError.h:138]: Excellent, thanks!

File: src/lib/core/CHIPError.h (hunk 2 above)
Comment (mwswartwout) [reply src/lib/core/CHIPError.h]: Good catch, fixed by the adding the loc argument to all instances of CHIP_INITIALIZE_ERROR_SOURCE

File: src/app/FailSafeContext.cpp [hunk 4]
```diff
@@ -85,12 +85,11 @@ void FailSafeContext::ScheduleFailSafeCleanup(FabricIndex fabricIndex, bool addN
 
     SetFailSafeArmed(false);
 
-    ChipDeviceEvent event;
-    event.Type                                                = DeviceEventType::kFailSafeTimerExpired;
-    event.FailSafeTimerExpired.fabricIndex                    = fabricIndex;
-    event.FailSafeTimerExpired.addNocCommandHasBeenInvoked    = addNocCommandInvoked;
-    event.FailSafeTimerExpired.updateNocCommandHasBeenInvoked = updateNocCommandInvoked;
-    CHIP_ERROR status                                         = PlatformMgr().PostEvent(&event);
+    ChipDeviceEvent event{ .Type                 = DeviceEventType::kFailSafeTimerExpired,
```
Comment (bzbarsky-apple) [inline src/app/FailSafeContext.cpp:88]: How are these changes related to this PR?

File: src/app/clusters/bindings/bindings.cpp [hunk 5]
```diff
@@ -259,9 +259,8 @@ CHIP_ERROR BindingTableAccess::WriteBindingTable(const ConcreteDataAttributePath
 
 CHIP_ERROR BindingTableAccess::NotifyBindingsChanged()
 {
-    DeviceLayer::ChipDeviceEvent event;
-    event.Type                        = DeviceLayer::DeviceEventType::kBindingsChangedViaCluster;
-    event.BindingsChanged.fabricIndex = mAccessingFabricIndex;
+    DeviceLayer::ChipDeviceEvent event{ .Type            = DeviceLayer::DeviceEventType::kBindingsChangedViaCluster,
```
Comment (bzbarsky-apple) [inline src/app/clusters/bindings/bindings.cpp:262]: How are these changes related to this PR?

File: src/include/platform/CHIPDeviceEvent.h [hunk 6]
```diff
@@ -530,7 +530,6 @@ struct ChipDeviceEvent final
         } OtaStateChanged;
     };
 
-    void Clear() { memset(this, 0, sizeof(*this)); }
```
Comment (bzbarsky-apple) [inline src/include/platform/CHIPDeviceEvent.h:533]: This seems unrelated?

**Review by bzbarsky-apple — APPROVED** [review]
Sorry for the lag here....

**Review by andy31415 — APPROVED** [review]
